stm32-modules/heater-shaker/include/heater-shaker/ack_cache.hpp

```diff
@@ -0,0 +1,69 @@
+/*
+** The ack cache is a way to keep around gcode messages that have not yet
+** been executed. It maps a copy of the actual gcode object sent by the
+** host with the internal message id.
+**
+** This is not done with an actual map because that would need to allocate.
+*/
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <variant>
+
+template <size_t max_size, typename... Contents>
+struct AckCache {
+    using Payload = std::variant<std::monostate, Contents...>;
+    AckCache() : cache{CacheWrapper{0, Payload(std::monostate())}} {}
+
+    static constexpr size_t size = max_size;
+
+    template <typename ContentElement>
+    auto add(const ContentElement& element) -> uint32_t {
+        for (auto cache_element = cache.begin(); cache_element != cache.end();
+             cache_element++) {
+            if (std::holds_alternative<std::monostate>(
+                    cache_element->contents)) {
+                cache_element->contents = Payload(element);
+                cache_element->id = next_id;
+                next_id++;
+                if (next_id == 0) {
+                    next_id++;
+                }
+                return cache_element->id;
+            }
+        }
+        return 0;
+    }
+
+    auto remove_if_present(uint32_t id) -> Payload {
+        auto which = std::find_if(
+            cache.begin(), cache.end(),
+            [&id](auto element) -> bool { return element.id == id; });
+        if (which == cache.end()) {
+            return Payload(std::monostate());
+        }
+        auto payload = which->contents;
+        which->contents = std::monostate();
+        which->id = 0;
+        return payload;
+    }
+
+    auto clear() -> void {
+        for (auto& cache_element : cache) {
+            cache_element.contents = std::monostate();
+            cache_element.id = 0;
+        }
+    }
+
+  private:
+    // Present only for testing; do not use
+    friend class _AckCacheTestHook;
```

Comment:
love the `friend class` name, and also did not realize cpp had a concept like this.

Reply:
yeah it can be pretty useful... although i consider it a pretty serious code smell in basically any other case than "this is for testing do not use it otherwise" since it indicates that somebody is either
a) just kind of giving up on finding appropriate usage restrictions
b) trying to implement extending classes or something bizarre